presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedSmokeTestBase.java

```diff
@@ -1031,6 +1031,30 @@ public void testCreateNestedPartitionedTable(Session session, FileFormat fileFor
         dropTable(session, "test_nested_table2");
     }
 
+    @Test
+    public void testPartitionedByTime()
+    {
+        testSelectOrPartitionedByTime(getSession(), false, FileFormat.PARQUET);
+        testSelectOrPartitionedByTime(getSession(), false, FileFormat.ORC);
+        testSelectOrPartitionedByTime(getSession(), true, FileFormat.PARQUET);
+        testSelectOrPartitionedByTime(getSession(), true, FileFormat.ORC);
+    }
+
+    private void testSelectOrPartitionedByTime(Session session, boolean partitioned, FileFormat format)
+    {
+        String tableName = format("test_%s_by_time", partitioned ? "partitioned" : "selected");
+        String partitioning = partitioned ? ", partitioning = ARRAY['x']" : "";
+        assertUpdate(format("CREATE TABLE %s (x TIME, y BIGINT) WITH (format = '%s'%s)", tableName, format, partitioning));
+        assertUpdate(format("INSERT INTO %s VALUES (TIME '10:12:34', 12345)", tableName), 1);
+        assertQuery(format("SELECT COUNT(*) FROM %s", tableName), "SELECT 1");
+        assertQuery(format("SELECT x FROM %s", tableName), "SELECT CAST('10:12:34' AS TIME)");
+        assertUpdate(format("INSERT INTO %s VALUES (TIME '9:00:00', 67890)", tableName), 1);
+        assertQuery(format("SELECT COUNT(*) FROM %s", tableName), "SELECT 2");
+        assertQuery(format("SELECT x FROM %s WHERE y = 12345", tableName), "SELECT CAST('10:12:34' AS TIME)");
+        assertQuery(format("SELECT x FROM %s WHERE y = 67890", tableName), "SELECT CAST('9:00:00' AS TIME)");
```

Comment:
Seems when executing filter by a column with `TimeType` on ORC, we will meet some problems, no matter the TimeType column is a partitioned column or non-partitioned column.

For reference, see test cases in `IcebergDistributedTestBase.testPartitionedByTimeType()`.

Reply:
Not sure what you mean... that test is also passing. Is there something I am missing from the test cases that I should add?

Reply:
The test `IcebergDistributedTestBase.testPartitionedByTimeType()` is currently running on `PARQUET` only, it includes the following scenarios that I think should also be added here to test on `ORC`:

```
assertQuery(format("SELECT y FROM %s WHERE x = time '10:12:34'", tableName), "values(12345)");
```

Reply:
Could you please expand on how you identified this issue? Looked through some of the orc code and I'm not sure what I'm missing. Thank you

Reply:
According to the Iceberg specification, the raw data for `TimeType` maintained in data files should be with microsecond precision. Meanwhile, the data for `TimeType` is maintained in presto with millisecond precision. So in order to support `TimeType` on PARQUET in PR #21337, we have done the following things:

1. Convert the data of type `TimeType` on writing, change the precision from milliseconds to microseconds. See: `TimeValueWriter`.

2. Convert the data of type `TimeType` on reading, change the precision from microseconds to milliseconds. See: `LongTimeMicrosColumnReader`.

3. Convert the `TimeType` values exist in filter predication. For Iceberg, filtering conditions need to be processed at two levels:
 - At the iceberg file plan level, see: `IcebergSplitManager.getSplits(...)` and `ExpressionConverter.getIcebergLiteralValue()`.
 - At the iceberg page source level, the predicate passed to the newly constructed `ConnectorPageSource` should be handled (for PARQUET, the filter has been completely ignored since `columnIndexFilterEnabled = false`), see: `IcebergPageSourceProvider.createParquetPageSource(...)`.

4. Handle the partitioned column with type `TimeType`, See `IcebergPageSink.getIcebergValue()`, `IcebergPageSource.nativeValueToBlock()`, and `PartitionTable.convert()`.

So based on this context, in order to support `TimeType` on ORC, we need to do the following file format specific things:

1. Convert the data of type `TimeType` on writing, change the precision from milliseconds to microseconds in `LongColumnWriter` and `LongDictionaryColumnWriter` etc.

2. Convert the data of type `TimeType` on reading, change the precision from microseconds to milliseconds in `LongDirectBatchStreamReader` and `LongDictionaryBatchStreamReader` etc.

3. Convert the `TimeType` value in filter predication at the iceberg page source level, convert or ignore the predicates on columns of type `TimeType` in `IcebergPageSourcePrivider.createBatchOrcPageSource()`.

In my local experiment, after doing all these things, the tests will all passing. You can have a try, and please let me know if there are any misunderstandings above.

Reply:
Still a little unclear on how to go about step 3... do you have a branch that you made your local changes on?

Reply:
Yes, you can refer to here: https://github.com/hantangwangd/presto/blob/support_timetype_on_orc/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java#L576-L585. In this code, I simply skipped the filtering predicate on `TimeType` column.

Reply:
Alright thanks. Was the error you got before making the fix the same as this?

```
java.lang.AssertionError: For query: 
 SELECT y FROM test_selected_by_time WHERE x = time '10:12:34'
 actual column types:
 [bigint]
expected column types:
[bigint]

not equal
Actual rows (1 of 1 extra rows shown, 2 rows in total):
    [54321]
```

I made the changes you requested and I am still getting this error on `testSelectOrPartitionedByTime`

Reply:
You have inserted another row, so the assertion statement should be:

```
assertQuery(format("SELECT y FROM %s WHERE x = time '10:12:34'", tableName), "values 12345, 54321");
```